File: 516. Longest Palindromic Subsequence/solution.py
class Solution(object):
    def longestPalindromeSubseq(self, s):
        """
        :type s: str
        :rtype: int
        """
        # edge case
        if not s:
            return 0
        
        # prepare for dp
        dp = []
        n = len(s)
        for i in range(n):
            dp.append([0]*n)
            
        # check for palindromic subsequences
        for i in range(n-1, -1, -1):
            dp[i][i] = 1
            for j in range(i + 1, n):
                if s[i] == s[j]:
                    dp[i][j] = 2 + dp[i+1][j-1]
                else:
                    dp[i][j] = max(dp[i+1][j], dp[i][j-1])
                    
        return dp[0][-1]

File: 516. Longest Palindromic Subsequence/solution.java
class Solution {
    public int longestPalindromeSubseq(String s) {
        if (s.length() < 2) {
            return s.length();
        }
        int[][] dp = new int[s.length()][s.length()];
        
        for (int i = 0; i < s.length(); i++) {
            dp[i][i] = 1;
        }
        
        for (int i = s.length(); i >= 0; i--) {
            for (int j = i + 1; j < s.length(); j++) {
                if (s.charAt(i) == s.charAt(j)) {
                    dp[i][j] = 2 + dp[i+1][j-1];
                } else {
                    dp[i][j] = Math.max(dp[i+1][j], dp[i][j-1]);
                }
            }
        }
        
        return dp[0][s.length()-1];
    }
}
